indexing-service/src/main/java/io/druid/indexing/overlord/MetadataTaskStorage.java

```diff
@@ -244,6 +244,75 @@ public void addLock(final String taskid, final TaskLock taskLock)
   }
 
   @Override
+  public void upgradeLock(String taskid, TaskLock taskLockToUpgrade)
+  {
+    Preconditions.checkNotNull(taskid, "taskid");
+    Preconditions.checkNotNull(taskLockToUpgrade, "taskLock");
+    int numUpgraded = 0;
+    boolean alreadyUpgraded = false;
+    final Map<Long, TaskLock> taskLocks = getLocksWithIds(taskid);
+
+    log.info("Found [%s] locks for task [%s] in TaskStorage",
+             taskLocks.size(),
+             taskid
+    );
+
+    for (final Map.Entry<Long, TaskLock> taskLockWithId : taskLocks.entrySet()) {
+      final long id = taskLockWithId.getKey();
+      final TaskLock taskLock = taskLockWithId.getValue();
+
+      // Candidate TaskLock will be equal to taskLockToUpgrade except it will be preemptive
+      // Checking equals for interval and version only as datasource, groupId and priority is common with the Task
+      if (taskLock.getInterval().equals(taskLockToUpgrade.getInterval()) &&
+          taskLock.getVersion().equals(taskLockToUpgrade.getVersion()) &&
+          !taskLock.isPreemptive().equals(taskLockToUpgrade.isPreemptive())) {
+        handler.updateLock(id, taskLockToUpgrade);
+        log.info("Upgraded lock id:[%s] on interval [%s] version [%s] for task:[%s] to non-preemptive lock",
+                 id,
+                 taskLockToUpgrade.getInterval(),
+                 taskLockToUpgrade.getVersion(),
+                 taskid
+        );
+        numUpgraded++;
+      } else if (taskLock.equals(taskLockToUpgrade)) {
+        alreadyUpgraded = true;
+      }
+    }
+
+    if (numUpgraded > 0 && !alreadyUpgraded) {
+      // This is the normal case, when everything is working fine numUpgraded should be 1
+      // If there are multiple lock upgrades then most probably overlord restarted or leader changed and it reacquired the same lock
+      log.info(
+          "Upgraded [%s] locks on interval [%s] for task:[%s] to non-preemptive lock",
+          numUpgraded,
+          taskLockToUpgrade.getInterval(),
+          taskid
+      );
+    } else if (numUpgraded > 0 && alreadyUpgraded){
+      // This can happen when the overlord failed or restarted again during the runtime of this task while there were multiple non-preemptive lock for a task
+      log.info("Upgraded [%s] locks on interval [%s] for task:[%s] to non-preemptive lock",
+               numUpgraded,
+               taskLockToUpgrade.getInterval(),
+               taskid)
+      ;
+
+    } else if(numUpgraded == 0 && alreadyUpgraded) {
+      log.info(
+          "Lock on interval [%s] for task:[%s] already upgraded to non-preemptive lock",
+          taskLockToUpgrade.getInterval(),
+          taskid
+      );
+    } else if(numUpgraded == 0 && !alreadyUpgraded){
+      log.error(
```

Comment:
This should be an exception, I think, otherwise callers have no way of knowing if they actually have an upgraded lock